src/passport-saml/saml-post-signing.ts

```diff
@@ -1,42 +1,19 @@
-import { SignedXml } from "xml-crypto";
-import * as algorithms from "./algorithms";
 import { SamlSigningOptions } from "./types";
+import { signXml } from "./xml";
 
 const authnRequestXPath =
   '/*[local-name(.)="AuthnRequest" and namespace-uri(.)="urn:oasis:names:tc:SAML:2.0:protocol"]';
 const issuerXPath =
   '/*[local-name(.)="Issuer" and namespace-uri(.)="urn:oasis:names:tc:SAML:2.0:assertion"]';
-const defaultTransforms = [
-  "http://www.w3.org/2000/09/xmldsig#enveloped-signature",
-  "http://www.w3.org/2001/10/xml-exc-c14n#",
-];
 
 export function signSamlPost(
   samlMessage: string,
   xpath: string,
   options: SamlSigningOptions
 ): string {
-  if (!samlMessage) throw new Error("samlMessage is required");
-  if (!xpath) throw new Error("xpath is required");
-  if (!options) {
-    options = {} as SamlSigningOptions;
-  }
-
-  if (options.privateKey == null) throw new Error("options.privateKey is required");
-
-  const transforms = options.xmlSignatureTransforms || defaultTransforms;
-  const sig = new SignedXml();
-  if (options.signatureAlgorithm) {
-    sig.signatureAlgorithm = algorithms.getSigningAlgorithm(options.signatureAlgorithm);
-  }
-  sig.addReference(xpath, transforms, algorithms.getDigestAlgorithm(options.digestAlgorithm));
```

Comment:
Note that this is different from the reference used bellow for location, but my factorization make it the same (so it becomes `sig.addReference(xpath + issuerXPath, transforms, algorithms.getDigestAlgorithm(options.digestAlgorithm));` )

I don't know if that is good, bad or indifferent, but I notice it doesn't break any tests :) Let me know if I should revert that change.

Reply:
I feel like this change could cause a problem, though I would need to do more testing. For example, in the tests "should sign a simple saml request" and "should place the Signature element after the Issuer element", the Issuer is the same, but the Request is different, so I would expect a different signature. After your change I see these two results for the tests:

```xml
<SAMLRequest>
	<saml2:Issuer
		xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion" Id="_0">http://example.com
	</saml2:Issuer>
	<Signature
		xmlns="http://www.w3.org/2000/09/xmldsig#">
		<SignedInfo>
			<CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
			<SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/>
			<Reference URI="#_0">
				<Transforms>
					<Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
					<Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				</Transforms>
				<DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
				<DigestValue>tmaZyo4Chzy/OjcBg5xxWCGNf1s=</DigestValue>
			</Reference>
		</SignedInfo>
		<SignatureValue>oXiXS9NooDNf7yBRx07P1+aMD2xktkyee/9gzBpdL+MTXsSZ7EBR1OfAM+biVwHaQoC0SW6RmyNYM5CSOAELVGwS9rS1tUB79RqrIgrR0O3GbSX1fsmTULB5Xne2s/8EnY8qPPIGhVcx3zCBW3zNzPwyFePj+U4oMOk52CqSu/mjVuU6pzrbbl4H9jYAg6S23kzJPZGVexfHDkU26wmQPnwisZ1TI9bGwkF2I+wxab2l33LJMnYBBKzPaIM6DxgAFRX+wSnTdGCqxiMcWHD+cr/Mcug8/169Lwaeflt4Bo9+2ojtfZkNtj8ggonR3cAA0KqGCEfAJYUguI5VCtfgyw==</SignatureValue>
	</Signature>
</SAMLRequest>
```

and

```xml
<SAMLRequest>
	<saml2:Issuer
		xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion" Id="_0">http://example.com
	</saml2:Issuer>
	<Signature
		xmlns="http://www.w3.org/2000/09/xmldsig#">
		<SignedInfo>
			<CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
			<SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/>
			<Reference URI="#_0">
				<Transforms>
					<Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
					<Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				</Transforms>
				<DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
				<DigestValue>tmaZyo4Chzy/OjcBg5xxWCGNf1s=</DigestValue>
			</Reference>
		</SignedInfo>
		<SignatureValue>oXiXS9NooDNf7yBRx07P1+aMD2xktkyee/9gzBpdL+MTXsSZ7EBR1OfAM+biVwHaQoC0SW6RmyNYM5CSOAELVGwS9rS1tUB79RqrIgrR0O3GbSX1fsmTULB5Xne2s/8EnY8qPPIGhVcx3zCBW3zNzPwyFePj+U4oMOk52CqSu/mjVuU6pzrbbl4H9jYAg6S23kzJPZGVexfHDkU26wmQPnwisZ1TI9bGwkF2I+wxab2l33LJMnYBBKzPaIM6DxgAFRX+wSnTdGCqxiMcWHD+cr/Mcug8/169Lwaeflt4Bo9+2ojtfZkNtj8ggonR3cAA0KqGCEfAJYUguI5VCtfgyw==</SignatureValue>
	</Signature>
	<SomeOtherElement/>
</SAMLRequest>
```

You'll note that they are identical except for the `<SomeOtherElement />`. Now, note before this change (literally one I just reverted the last commit you made on your branch):

```xml
<SAMLRequest Id="_0">
	<saml2:Issuer
		xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">http://example.com
	</saml2:Issuer>
	<Signature
		xmlns="http://www.w3.org/2000/09/xmldsig#">
		<SignedInfo>
			<CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
			<SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/>
			<Reference URI="#_0">
				<Transforms>
					<Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
					<Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				</Transforms>
				<DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
				<DigestValue>1yis05FW/NgGxi12sn/bW3GP9co=</DigestValue>
			</Reference>
		</SignedInfo>
		<SignatureValue>Oa5ST39rynnUH6XN4tnjoK2luRlKGOq4VHPAKqSgEjzEymTFQRhMwqwQTuFI+AwHSn0qd4wc7GGLIHn0BmUsk/CBZ51nvgjiyTQo+Gkc2/24QlCAwpOM35hgOEaMMvJXgzkFwxvnV/3TGA2J+jrrcQ0q2l6nSuDe27JnCCzbo1vFiHIuWG91pZnS0ZQKnJ593jG5ozo2m2a7l/KvCXIWCGs91KR43IKgmQmOIkVk4i170Ep2trlyj5651LFlT4LShDkkrf4tvWAmeC7rZgf97j58m9vTYXY7zZt5URIvmlE9SZH6NmUdrryZjfZin4Xf7FqpfK/sLzVfBCSLvCse8A==</SignatureValue>
	</Signature>
</SAMLRequest>
```

and

```xml
<SAMLRequest Id="_0">
	<saml2:Issuer
		xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">http://example.com
	</saml2:Issuer>
	<Signature
		xmlns="http://www.w3.org/2000/09/xmldsig#">
		<SignedInfo>
			<CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
			<SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/>
			<Reference URI="#_0">
				<Transforms>
					<Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
					<Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				</Transforms>
				<DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
				<DigestValue>5z/cj0WrzI7UGvPSu543FzezwE4=</DigestValue>
			</Reference>
		</SignedInfo>
		<SignatureValue>ggJj62O7r2KwAFyiRtQmWOGu2q1MtLFNcGGCYaeKi5/lmSwhjOqgrLLQlZnQWRIoxhkKw0OLgtT9dIeZC3+TMZtolLQ9OM0pQARz8svJuYQUd4ti71hoIRTRzgzEXbOvpyDoqXaJMeZeveidAg/DHIIATpCUwqy1soUPxiHXdweXJ8BYrNoWjFBKLULbBNmTlYEdeQqYOE3TcCuvCDYOdbQlTKRPJlC5eVz+SIRb8q6c9sHTzmFuiquFezQsroY3MEvHjO5jPQ1L3L1drVwWBeF8TNoNsHcJ8aMMMJqHzjiSDeAJZnq6G3VOisZrJVcSJWipAd383EyuhDy3Zo5PJw==</SignatureValue>
	</Signature>
	<SomeOtherElement/>
</SAMLRequest>
```

Here you'll note that the signature is different. So, this change that you made results in a signature that doesn't actually sign the contents of the Response, but just the Issuer, which means that any change we want can be made to the response and the signature would still check.

So, we need to do two things:

1. Put in a new test to catch this case
2. Make sure that the code changes that your making don't cause this problem

Reply:
I had figured out that something like this could be happening, that makes sense now. I'll add the test case and fix it 👍

Reply:
should be better now